lib/bap/bap.mli

```diff
@@ -3822,6 +3860,377 @@ module Std : sig
+        same meaning as in {!after}.  *)
+    val before : ('a,'b) cls -> ?rev:bool -> 'a term -> tid -> 'b term seq
+
+    (** [append t ~after:this p c] returns the [p] term with [c]
+        appended after [this] term. If [after] is not specified,
+        then append [def] to the end of the parent term (if it makes
+        sense, otherwise it is just added).  If [this] doesn't
+        occur in the [p] term then do nothing. The term tid is
+        preserved.  *)
+    val append : ('a,'b) cls -> ?after:tid -> 'a term -> 'b term -> 'a term
+
+    (** [prepend t ~before:this p c] returns the [p] with [c] inserted
+        before [this] term. If [before] is left unspecified, then
+        insert the [c] at the beginning of the [p] if it is a sequence,
+        otherwise just insert. If [this] is specified but doesn't occur
+        in the [p] then [p] is returned as is.  In all cases, the
+        returned term has the same [tid] as [p]. *)
+    val prepend : ('a,'b) cls -> ?before:tid -> 'a term -> 'b term -> 'a term
+
+  end
+
+  (** Program.  *)
+  module Program : sig
+    (** Program is a collection of function terms. *)
+
+    type t = program term
+
+    (** [create ()] creates an empty program. *)
+    val create : unit -> t
+
+    (** [lookup t program id] is like {{!find}find} but performs deep
+        lookup in the whole [program] for a term with a given [id].
+        This function is memoized, so it has amortized O(1) complexity,
+        with a wostcase complexity of $O(N)$, where $N$ is the total
+        amount of terms in program.  *)
+    val lookup : (_,'b) cls -> t -> tid -> 'b term option
+
+    (** [parent t program id] is [Some p] iff [find t p id <> None]  *)
+    val parent : ('a,'b) cls -> t -> tid -> 'a term option
+
+    module Builder : sig
+      type t
+      val create : ?tid:tid  -> ?subs:int -> unit -> t
+      val add_sub : t -> sub term -> unit
+      val result : t -> program term
+    end
+
+    include Regular with type t := t
+  end
+
+  (** Function.  *)
+  module Sub : sig
+    (** Function is a set of blocks.
+        The first block of a function is considered an entry block.
+    *)
+    type t = sub term
+
+    val create : ?name:string -> unit -> t
+
+    val name : t -> string
+
+    module Builder : sig
+      type t
+      val create : ?tid:tid -> ?args:int -> ?blks:int -> ?name:string -> unit -> t
+      val add_blk : t -> blk term -> unit
+      val add_arg : t -> arg term -> unit
+      val result : t -> sub term
+    end
+
+    include Regular with type t := t
+  end
+
+  (** Block.  *)
+  module Blk : sig
+    (** Logically block consists of a set of {{!Phi}phi nodes}, a
+        sequence of {{!Def}definitions} and a sequence of out-coming
+        edges, aka {{!Jmp}jumps}. A colloquial term for this three
+        entities is a "block element".
+
+        The order of $\Phi$-nodes is unspecified. Definitions are stored
```

Comment:
Be explicit on why the order is unspecified, e.g., The order of $\Phi$-nodes is left unspecified because the order does not affect the execution semantics.